---
title: "easier_manual"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{easier_manual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(easier)
```

---

# Introduction {#introduction} 

This vignette describes how to use the `easier` package for the prediction and interpretation of cancer patients' systems immune response. 

This packages provides a Shiny application to allow users to finely explore patient-specific systems biomarkers of the immune response, including composition of the immune repertoire, and activity of intra- and inter- cellular communications.

To gain some insight into the fundamentals of `easier` package, we recommend reading the following preprint [article](https://www.biorxiv.org/content/10.1101/2021.02.05.429977v1).

In order to use `r BiocStyle::Biocpkg("easier")` in your analysis, bulk-tumor RNA sequencing data is required as input (when available, patient response to immunotherapy can be optionally entered):

- `RNA_counts`, a `data.frame` containing TPM values with HGNC symbols in rows and samples in columns. 
- `RNA_tpm`, a `data.frame` containing raw counts values with HGNC symbols in rows and samples in columns. 
- `real_patient_response`, a character `vector` containing clinical patients' response (where non-responders are labeled as NR and responders as R).

In this vignette, we will illustrate the main features of `easier` on a publicly available melanoma dataset including both response to ICB therapy and RNA sequencing from Gide, et al. "Distinct Immune Cell Populations Define Response to Anti-PD-1 Monotherapy and Anti-PD-1/Anti-CTLA-4 Combined Therapy", published in Cancer Cell, 2019 [@Gide2019]
[doi:10.1016/j.ccell.2019.01.003](https://doi.org/10.1016/j.ccell.2019.01.003). ** Still need to ask for consent.

The RNA sequencing fastq files are available from NCBI BioProject [PRJEB23709](https://www.ncbi.nlm.nih.gov/bioproject/522397).

In this dataset, we considered 41 metastatic melanoma patients () treated with anti-PD1 with prior BRAFi therapy.

# Getting started {#gettingstarted}

Starting R, this package can be installed as follows:

```{r eval = FALSE}
# install.packages("~/ownCloud/SystemsImmunoOncology/Mechanistic_signatures_project/easier_1.0.3.tar.gz", repos = NULL, type="source")
library(easier)
````

If you have your bulk-tumor RNA sequencing data ready, you can compute five systems-based signatures of the tumor microenvironment by running:


```{r eval = FALSE}
# Example: Riaz
data("Riaz_data")

# Computation of cell fractions
cell_fractions <- compute_cell_fractions(RNA_tpm = Riaz_data$tpm_RNAseq)

# Computation of pathway scores
pathway_activity <- compute_pathways_scores(RNA_counts = Riaz_data$raw_counts_RNAseq,
                                            remove_genes_ICB_proxies = TRUE)

# Computation of TF activity
tf_activity <- compute_TF_activity(RNA_tpm = Riaz_data$tpm_RNAseq, 
                                   remove_genes_ICB_proxies = FALSE)

# Computation of ligand-receptor pair weights
lrpair_weights <- compute_LR_pairs(RNA_tpm = Riaz_data$tpm_RNAseq, 
                                   remove_genes_ICB_proxies = FALSE, 
                                   cancertype = "pancan")

# Computation of cell-cell interaction scores
ccpair_scores <- compute_CC_pairs_grouped(lrpairs = lrpair_weights, 
                                          cancertype = "pancan")
````

You can additionally compute published hallmarks of the immune response on your patients' data. 

Hallmark of the immune response | Original study |
------------------------------- | -------------- |
Cytolytic activity (CYT) | Rooney et al, Cell, 2015 |
Roh immune score (Roh_IS) | Roh et al., Sci. Transl. Med., 2017 |
Chemokine signature (chemokines) | Messina et al., Nat. Sci. Rep., 2012 |
Davoli immune signature (Davoli_IS) | Davoli et al., Science 2017 |
IFNy signature (IFNy) | Ayers et al., JCI, 2017 |
Expanded immune signature (Ayers_expIS) | Ayers et al., JCI, 2017 |
T-cell inflamed signature (Tcell_inflamed) | Ayers et al., JCI, 2017 |
Repressed immune resistance (RIR) | Jerby-Arnon et al., Cell, 2018 |
Tertiary lymphoid structures signature (TLS) | Cabrita et al., Nature, 2020 |

By default all hallmarks of the immune response are used, however you can set your own `list_gold_standards`.

```{r eval = FALSE}
tasks <- c("CYT", "Roh_IS", "chemokines", "Davoli_IS", "IFNy", "Ayers_expIS", "Tcell_inflamed", "RIR", "TLS")
immune_response <- compute_gold_standards(RNA_tpm = Riaz_data$tpm_RNAseq, 
                                          list_gold_standards = tasks, 
                                          cancertype = "SKCM",
                                          output_file_path = "/Users/Oscar/Desktop/Riaz")
```

We are going to use the systems-based signatures computed previously to predict patients' immune response. `predict_immune_response` returns compute predictions for single system-based signatures and pairwise combinations of these system-based signatures when `include_pairwise_combos = TRUE`. 

Because models were built in a cancer-type-specific fashion, the user is required to indicate to which `cancertype` the bulk RNA-seq data belongs.

```{r eval = FALSE}
predictions_immune_response <- predict_immune_response(pathways = pathway_activity,
                                                       immunecells = cell_fractions,
                                                       lrpairs = lrpair_weights,
                                                       tfs = tf_activity,
                                                       ccpairs = ccpair_scores,
                                                       include_pairwise_combos = TRUE,
                                                       cancertype = "SKCM")
```

Depending on the patient's response information available, we can now think of two possible scenarios where:

- `real_patient_response` is **known** and the performance of the computed predictions can be evaluated. In this case, we expect the user to provide a character `vector` containing clinical patients' response (where non-responders are labeled as NR and responders as R).
- `real_patient_response` is **unknown** and an score is provided assessing patient-specific likelihood of response to therapy.


```{r eval = FALSE}
compare_immune_response(predictions_immune_response = predictions_immune_response,
                        real_patient_response = Riaz_data$patient_response,
                        RNA_tpm = Riaz_data$tpm_RNAseq,
                        output_file_path = "/Users/Oscar/Desktop/Riaz",
                        cancertype = "SKCM")
```

We can go further and analyze systems biomarkers present in the user's bulk RNA-seq data. `explore_biomarkers` allows to investigate mechanisms behind patients' response to treatment through interpretable biomarkers. 

In order to leverage the biomarkers information obtained during model training, you need to specify again which `cancertype` the bulk RNA-seq data belongs.

```{r eval = FALSE}
explore_biomarkers(pathways = pathways_scores,
                   immunecells = cell_fractions,
                   # lrpairs = lrpairs_weights,
                   # tfs = tfs_scores,
                   # ccpairsgroupedscores = ccpairsgrouped_scores,
                   cancertype = cancertype,
                   real_patient_response = Riaz_data$patient_response,
                   output_file_path = "/Users/Oscar/Desktop/Riaz")
```

## Ready to go!

Once you go through the full framework of `easier`, you are ready to engage in the app of `easier`. To do so, you can just launch the `easier()` app:


# Session info {-}

```{r}
sessionInfo()
```

