---
title: "easier User Manual"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{easier_manual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(easier)
```

---

# Introduction {#introduction} 

This vignette describes how to use the `easier` package for streamlining your workflow around the `easier` tool. `easier` is intended to assist researchers in assessing patients' likelihood
to respond to ICB therapies providing just the patients' RNA-seq data as input.

This packages provides a Shiny application to allow users to finely explore patient-specific systems biomarkers of immune response inspecting the tumor microenvironment from several points of view, including composition of the tumor immune contexture, and activity of intra- and extra-cellular communication.

To gain some insight into the fundamentals of `easier` package, we recommend reading our original work: Lapuente-Santana et al. "Predictive systems biomarkers of response to immune checkpoint inhibitors", available in bioRxiv [doi:10.1101/2021.02.05.429977](https://www.biorxiv.org/content/10.1101/2021.02.05.429977v1).

In order to use `easier` in your study, bulk-tumor RNA sequencing data is required as input (when available, patients' response to immunotherapy can be optionally provided):

- `RNA_counts`, a `data.frame` containing raw counts values with HGNC symbols in rows and samples in columns. 
- `RNA_tpm`, a `data.frame` containing TPM values with HGNC symbols in rows and samples in columns. 
- `real_patient_response`, a character `vector` containing clinical patients' response (where non-responders are labeled as NR and responders as R).

Along this vignette we will illustrate the main features of `easier` on a publicly available bladder cancer dataset including both response to ICB therapy and RNA sequencing from Mariathasan et al. "TGF-B attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells", published in Nature, 2018 [doi:10.1038/nature25501](https://doi.org/10.1038/nature25501). The processed data is made available via [`IMvigor210CoreBiologies`](http://research-pub.gene.com/IMvigor210CoreBiologies/) package under the CC-BY license.

Here we will consider bladder cancer patients with complete response (CR) as responders and patients with progressive disease (PD) as non-responders.

# Getting started {#gettingstarted}

Starting R, this package can be installed as follows:

```{r, eval=FALSE}
install.packages("../easier_0.9.0.tar.gz", repos = NULL, type="source")
library(easier)
````

We first load the processed data provided in IMvigor210CoreBiologies from Mariathasan cohort as follows:

```{r, eval=FALSE}
data("dataset_mariathasan")
````

If you have your bulk-tumor RNA sequencing data ready, you can compute the five quantitative descriptors of the tumor microenvironment described below.

| Quantitative descriptor  | Descriptor conception  | Prior knowledge |
|------------------------ | ----------------------------------- | ----------------------------------- |
| Pathway activity | Holland et al., BBAGRM, 2019; Schubert et al., Nat Commun, 2018 | Holland et al., BBAGRM, 2019; Schubert et al., Nat Commun, 2018 |
| Immune cell quantification | Finotello et al., Genome Med, 2019 | Finotello et al., Genome Med, 2019|
| TF activity | Garcia-Alonso et al., Genome Res, 2019 | Garcia-Alonso et al., Genome Res, 2019 |
| Ligand-Receptor pairs | Lapuente-Santana et al., bioRxiv, 2021 | Ramilowski et al., Nat Commun, 2015 |
| Cell-cell interaction | Lapuente-Santana et al., bioRxiv, 2021 | Ramilowski et al., Nat Commun, 2015 |

Here we use RNA-sequencing data from Mariathasan cohort to compute the five quantitative descriptors of the tumor microenvironment.

```{r eval = FALSE}
# Obtain cell fractions
cell_fractions <- compute_cell_fractions(RNA_tpm = dataset_mariathasan@tpm)

# Obtain pathway scores
pathway_activity <- compute_pathways_scores(RNA_counts = dataset_mariathasan@counts,
                                            remove_genes_ICB_proxies = TRUE)

# Obtain TF activity
tf_activity <- compute_TF_activity(RNA_tpm = dataset_mariathasan@tpm, 
                                   remove_genes_ICB_proxies = FALSE)

# Obtain ligand-receptor pair weights
lrpair_weights <- compute_LR_pairs(RNA_tpm = dataset_mariathasan@tpm, 
                                   remove_genes_ICB_proxies = FALSE, 
                                   cancer_type = "pancan")

# Obtain cell-cell interaction scores
ccpair_scores <- compute_CC_pairs_grouped(lrpairs = lrpair_weights, 
                                          cancer_type = "pancan")
````

You can additionally compute published hallmarks of the immune response on your patients' data. 

Hallmark of the immune response | Original study |
------------------------------- | -------------- |
Cytolytic activity (CYT) | Rooney et al, Cell, 2015 |
Roh immune score (Roh_IS) | Roh et al., Sci. Transl. Med., 2017 |
Chemokine signature (chemokines) | Messina et al., Nat. Sci. Rep., 2012 |
Davoli immune signature (Davoli_IS) | Davoli et al., Science 2017 |
IFNy signature (IFNy) | Ayers et al., JCI, 2017 |
Expanded immune signature (Ayers_expIS) | Ayers et al., JCI, 2017 |
T-cell inflamed signature (Tcell_inflamed) | Ayers et al., JCI, 2017 |
Repressed immune resistance (RIR) | Jerby-Arnon et al., Cell, 2018 |
Tertiary lymphoid structures signature (TLS) | Cabrita et al., Nature, 2020 |
Immuno-predictive score (IMPRES) | Auslander et al., Nat.Med., 2018 |
Microsatellite instability status | Fu et al., BMC Genomics, 2019 |

By default nine hallmarks of the immune response are used (see below), however you can set your own preferences by modifying the argument `list_gold_standards`.

```{r eval = FALSE}
tasks <- c("CYT", "Roh_IS", "chemokines", "Davoli_IS", "IFNy", "Ayers_expIS", "Tcell_inflamed", "RIR", "TLS")
tasks_values <- compute_gold_standards(RNA_tpm = dataset_mariathasan@tpm, 
                                       list_gold_standards = tasks)
```

We are going to use the quantitative descriptors. computed previously to predict patients' immune response. `predict_immune_response` returns compute predictions for each quantitative descriptor.

Because models were built in a cancer-type-specific fashion, the user is required to indicate which `cancer_type` learned model should be used. 

```{r eval = FALSE}

predictions_immune_response <- predict_immune_response(pathways = pathway_activity,
                                                       immunecells = cell_fractions,
                                                       tfs = tf_activity,
                                                       lrpairs = lrpair_weights,
                                                       ccpairs = ccpair_scores,
                                                       cancer_type = dataset_mariathasan@cancertype, 
                                                       verbose = TRUE)

```

Depending on the patient's response information available, we can now think of two possible scenarios where:

- `real_patient_response` is **known** and the performance of the computed predictions can be evaluated. In this case, we expect the user to provide a character `vector` containing clinical patients' response (where non-responders are labeled as NR and responders as R).
- `real_patient_response` is **unknown** and an score is provided assessing patient-specific likelihood of response to therapy.

For the first scenario, we can use the patients' response information available from Mariathasan cohort. Because patients' tumor mutational burden is also provided, we can also used to compare its performance against the quantitative descriptors. 

Additionally the user can set `easier_with_TMB = TRUE` to apply a refined approach based on the integration of `easier` predictions with information on tumor mutational burden (TMB). 
Since both immune response and TMB are essential for an effective immunotherapy response, we decided to conceptualize this in our predictions by either penalizing or weighting differently our scores in high- and low-TMB patients. 

```{r eval = FALSE}
# assess easier predicted immune response
assess_immune_response(predictions_immune_response = predictions_immune_response,
                       real_patient_response = dataset_mariathasan@response,
                       RNA_tpm = dataset_mariathasan@tpm,
                       output_file_path = "../figures",
                       cancer_type = dataset_mariathasan@cancertype,
                       TMB_values = dataset_mariathasan@TMB,
                       easier_with_TMB = TRUE)

```

We can go further and analyze systems biomarkers present in the user's bulk RNA-seq data. `explore_biomarkers` allows to investigate mechanisms behind patients' response to treatment through interpretable biomarkers. 

In order to leverage the biomarkers information obtained during model training, you need to specify again which `cancer_type` the bulk RNA-seq data belongs.

```{r eval = FALSE}
explore_biomarkers(pathways = pathway_activity,
                   immunecells = cell_fractions,
                   lrpairs = lrpair_weights,
                   tfs = tf_activity,
                   ccpairs = ccpair_scores,
                   cancer_type = dataset_mariathasan@cancertype,
                   real_patient_response = dataset_mariathasan@response,
                   output_file_path = "../figures",
                   TMB_values = dataset_mariathasan@TMB)
```

Moving to a real-case scenario where the patients' response is not known, making treatments decision is a must and `easier` aims to help clinicians in this regard.

**Think about unique patient score**


## Ready to go!

Once you go through the full framework of `easier`, you are ready to engage in the app of `easier`. To do so, you can just launch the `easier()` app:


# Session info {-}

```{r}
sessionInfo()
```

