<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>easier_manual</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">easier_manual</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(easier)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="co">#&gt; Loading required package: progeny</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co">#&gt; Loading required package: viper</span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="co">#&gt; Loading required package: Biobase</span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="co">#&gt; Loading required package: parallel</span></span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="co">#&gt; </span></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co">#&gt; Attaching package: &#39;BiocGenerics&#39;</span></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="co">#&gt; The following objects are masked from &#39;package:parallel&#39;:</span></span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="co">#&gt; </span></span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span></span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="co">#&gt;     clusterExport, clusterMap, parApply, parCapply, parLapply,</span></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="co">#&gt;     parLapplyLB, parRapply, parSapply, parSapplyLB</span></span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="co">#&gt; </span></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-18"><a href="#cb1-18"></a><span class="co">#&gt; </span></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span></span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span></span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span></span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="co">#&gt;     union, unique, unsplit, which, which.max, which.min</span></span>
<span id="cb1-25"><a href="#cb1-25"></a><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span id="cb1-26"><a href="#cb1-26"></a><span class="co">#&gt; </span></span>
<span id="cb1-27"><a href="#cb1-27"></a><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span id="cb1-28"><a href="#cb1-28"></a><span class="co">#&gt;     &#39;browseVignettes()&#39;. To cite Bioconductor, see</span></span>
<span id="cb1-29"><a href="#cb1-29"></a><span class="co">#&gt;     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</span></span>
<span id="cb1-30"><a href="#cb1-30"></a><span class="co">#&gt; Loading required package: DESeq2</span></span>
<span id="cb1-31"><a href="#cb1-31"></a><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span id="cb1-32"><a href="#cb1-32"></a><span class="co">#&gt; Loading required package: stats4</span></span>
<span id="cb1-33"><a href="#cb1-33"></a><span class="co">#&gt; </span></span>
<span id="cb1-34"><a href="#cb1-34"></a><span class="co">#&gt; Attaching package: &#39;S4Vectors&#39;</span></span>
<span id="cb1-35"><a href="#cb1-35"></a><span class="co">#&gt; The following object is masked from &#39;package:base&#39;:</span></span>
<span id="cb1-36"><a href="#cb1-36"></a><span class="co">#&gt; </span></span>
<span id="cb1-37"><a href="#cb1-37"></a><span class="co">#&gt;     expand.grid</span></span>
<span id="cb1-38"><a href="#cb1-38"></a><span class="co">#&gt; Loading required package: IRanges</span></span>
<span id="cb1-39"><a href="#cb1-39"></a><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span id="cb1-40"><a href="#cb1-40"></a><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span id="cb1-41"><a href="#cb1-41"></a><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span id="cb1-42"><a href="#cb1-42"></a><span class="co">#&gt; Loading required package: DelayedArray</span></span>
<span id="cb1-43"><a href="#cb1-43"></a><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span id="cb1-44"><a href="#cb1-44"></a><span class="co">#&gt; </span></span>
<span id="cb1-45"><a href="#cb1-45"></a><span class="co">#&gt; Attaching package: &#39;matrixStats&#39;</span></span>
<span id="cb1-46"><a href="#cb1-46"></a><span class="co">#&gt; The following objects are masked from &#39;package:Biobase&#39;:</span></span>
<span id="cb1-47"><a href="#cb1-47"></a><span class="co">#&gt; </span></span>
<span id="cb1-48"><a href="#cb1-48"></a><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span id="cb1-49"><a href="#cb1-49"></a><span class="co">#&gt; </span></span>
<span id="cb1-50"><a href="#cb1-50"></a><span class="co">#&gt; Attaching package: &#39;DelayedArray&#39;</span></span>
<span id="cb1-51"><a href="#cb1-51"></a><span class="co">#&gt; The following objects are masked from &#39;package:matrixStats&#39;:</span></span>
<span id="cb1-52"><a href="#cb1-52"></a><span class="co">#&gt; </span></span>
<span id="cb1-53"><a href="#cb1-53"></a><span class="co">#&gt;     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></span>
<span id="cb1-54"><a href="#cb1-54"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-55"><a href="#cb1-55"></a><span class="co">#&gt; </span></span>
<span id="cb1-56"><a href="#cb1-56"></a><span class="co">#&gt;     aperm, apply, rowsum</span></span>
<span id="cb1-57"><a href="#cb1-57"></a><span class="co">#&gt; Loading required package: ROCR</span></span>
<span id="cb1-58"><a href="#cb1-58"></a><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span id="cb1-59"><a href="#cb1-59"></a><span class="co">#&gt; Loading required package: dorothea</span></span>
<span id="cb1-60"><a href="#cb1-60"></a><span class="co">#&gt; Loading required package: preprocessCore</span></span>
<span id="cb1-61"><a href="#cb1-61"></a><span class="co">#&gt; Loading required package: pdist</span></span>
<span id="cb1-62"><a href="#cb1-62"></a><span class="co">#&gt; Loading required package: arules</span></span>
<span id="cb1-63"><a href="#cb1-63"></a><span class="co">#&gt; Loading required package: Matrix</span></span>
<span id="cb1-64"><a href="#cb1-64"></a><span class="co">#&gt; </span></span>
<span id="cb1-65"><a href="#cb1-65"></a><span class="co">#&gt; Attaching package: &#39;Matrix&#39;</span></span>
<span id="cb1-66"><a href="#cb1-66"></a><span class="co">#&gt; The following object is masked from &#39;package:S4Vectors&#39;:</span></span>
<span id="cb1-67"><a href="#cb1-67"></a><span class="co">#&gt; </span></span>
<span id="cb1-68"><a href="#cb1-68"></a><span class="co">#&gt;     expand</span></span>
<span id="cb1-69"><a href="#cb1-69"></a><span class="co">#&gt; </span></span>
<span id="cb1-70"><a href="#cb1-70"></a><span class="co">#&gt; Attaching package: &#39;arules&#39;</span></span>
<span id="cb1-71"><a href="#cb1-71"></a><span class="co">#&gt; The following object is masked from &#39;package:SummarizedExperiment&#39;:</span></span>
<span id="cb1-72"><a href="#cb1-72"></a><span class="co">#&gt; </span></span>
<span id="cb1-73"><a href="#cb1-73"></a><span class="co">#&gt;     coverage</span></span>
<span id="cb1-74"><a href="#cb1-74"></a><span class="co">#&gt; The following object is masked from &#39;package:GenomicRanges&#39;:</span></span>
<span id="cb1-75"><a href="#cb1-75"></a><span class="co">#&gt; </span></span>
<span id="cb1-76"><a href="#cb1-76"></a><span class="co">#&gt;     coverage</span></span>
<span id="cb1-77"><a href="#cb1-77"></a><span class="co">#&gt; The following object is masked from &#39;package:IRanges&#39;:</span></span>
<span id="cb1-78"><a href="#cb1-78"></a><span class="co">#&gt; </span></span>
<span id="cb1-79"><a href="#cb1-79"></a><span class="co">#&gt;     coverage</span></span>
<span id="cb1-80"><a href="#cb1-80"></a><span class="co">#&gt; The following object is masked from &#39;package:S4Vectors&#39;:</span></span>
<span id="cb1-81"><a href="#cb1-81"></a><span class="co">#&gt; </span></span>
<span id="cb1-82"><a href="#cb1-82"></a><span class="co">#&gt;     decode</span></span>
<span id="cb1-83"><a href="#cb1-83"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb1-84"><a href="#cb1-84"></a><span class="co">#&gt; </span></span>
<span id="cb1-85"><a href="#cb1-85"></a><span class="co">#&gt;     abbreviate, write</span></span></code></pre></div>
<hr />
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This vignette describes how to use the <code>easier</code> package for the prediction and interpretation of cancer patients’ systems immune response.</p>
<p>This packages provides a Shiny application to allow users to finely explore patient-specific systems biomarkers of immune response inspecting the tumor microenvironment from several points of view, including composition of the tumor immune contexture, and activity of intra- and extra-cellular communication.</p>
<p>To gain some insight into the fundamentals of <code>easier</code> package, we recommend reading our original work: Lapuente-Santana et al. “Predictive systems biomarkers of response to immune checkpoint inhibitors”, available in bioRxiv <a href="https://www.biorxiv.org/content/10.1101/2021.02.05.429977v1">doi:10.1101/2021.02.05.429977</a>.</p>
<p>In order to use <code>easier</code> in your study, bulk-tumor RNA sequencing data is required as input (when available, patient response to immunotherapy can be optionally entered):</p>
<ul>
<li><code>RNA_counts</code>, a <code>data.frame</code> containing raw counts values with HGNC symbols in rows and samples in columns.</li>
<li><code>RNA_tpm</code>, a <code>data.frame</code> containing TPM values with HGNC symbols in rows and samples in columns.</li>
<li><code>real_patient_response</code>, a character <code>vector</code> containing clinical patients’ response (where non-responders are labeled as NR and responders as R).</li>
</ul>
<p>In this vignette, we will illustrate the main features of <code>easier</code> on a publicly available bladder cancer dataset including both response to ICB therapy and RNA sequencing from Mariathasan et al. “TGF-B attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells”, published in Nature, 2018 <a href="https://doi.org/10.1038/nature25501">doi:10.1038/nature25501</a>.</p>
<p>RNA sequencing data of Mariathasan cohort is available through <code>IMvigor210CoreBiologies</code> package which can be downloaded from <a href="http://research-pub.gene.com/IMvigor210CoreBiologies/IMvigor210CoreBiologies.tar.gz" class="uri">http://research-pub.gene.com/IMvigor210CoreBiologies/IMvigor210CoreBiologies.tar.gz</a>. Please install and load <code>IMvigor210CoreBiologies</code> as indicated so the data from Mariathasan cohort is accessible.</p>
<p>For the example dataset, we considered bladder cancer patients with complete response (CR) as responders and patients with progressive disease as non-responders.</p>
</div>
<div id="gettingstarted" class="section level1">
<h1>Getting started</h1>
<p>Starting R, this package can be installed as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">install.packages</span>(<span class="st">&quot;/Users/Oscar/ownCloud2/SystemsImmunoOncology/easier_project/easier_0.9.0.tar.gz&quot;</span>, <span class="dt">repos =</span> <span class="ot">NULL</span>, <span class="dt">type=</span><span class="st">&quot;source&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">library</span>(easier)</span></code></pre></div>
<p>If you have your bulk-tumor RNA sequencing data ready, you can compute five quantitative descriptors of the tumor microenvironment.</p>
<table>
<colgroup>
<col width="25%"></col>
<col width="37%"></col>
<col width="37%"></col>
</colgroup>
<thead>
<tr class="header">
<th>Quantitative descriptor</th>
<th>Descriptor conception</th>
<th>Prior knowledge</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Pathway activity</td>
<td>Holland et al., BBAGRM, 2019; Schubert et al., Nat Commun, 2018</td>
<td>Holland et al., BBAGRM, 2019; Schubert et al., Nat Commun, 2018</td>
</tr>
<tr class="even">
<td>Immune cell quantification</td>
<td>Finotello et al., Genome Med, 2019</td>
<td>Finotello et al., Genome Med, 2019</td>
</tr>
<tr class="odd">
<td>TF activity</td>
<td>Garcia-Alonso et al., Genome Res, 2019</td>
<td>Garcia-Alonso et al., Genome Res, 2019</td>
</tr>
<tr class="even">
<td>Ligand-Receptor pairs</td>
<td>Lapuente-Santana et al., bioRxiv, 2021</td>
<td>Ramilowski et al., Nat Commun, 2015</td>
</tr>
<tr class="odd">
<td>Cell-cell interaction</td>
<td>Lapuente-Santana et al., bioRxiv, 2021</td>
<td>Ramilowski et al., Nat Commun, 2015</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co"># Example: Riaz</span></span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="kw">data</span>(<span class="st">&quot;Riaz_data&quot;</span>)</span>
<span id="cb3-3"><a href="#cb3-3"></a></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co"># Computation of cell fractions</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>cell_fractions &lt;-<span class="st"> </span><span class="kw">compute_cell_fractions</span>(<span class="dt">RNA_tpm =</span> Riaz_data<span class="op">$</span>tpm_RNAseq)</span>
<span id="cb3-6"><a href="#cb3-6"></a></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co"># Computation of pathway scores</span></span>
<span id="cb3-8"><a href="#cb3-8"></a>pathway_activity &lt;-<span class="st"> </span><span class="kw">compute_pathways_scores</span>(<span class="dt">RNA_counts =</span> Riaz_data<span class="op">$</span>raw_counts_RNAseq,</span>
<span id="cb3-9"><a href="#cb3-9"></a>                                            <span class="dt">remove_genes_ICB_proxies =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-10"><a href="#cb3-10"></a></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co"># Computation of TF activity</span></span>
<span id="cb3-12"><a href="#cb3-12"></a>tf_activity &lt;-<span class="st"> </span><span class="kw">compute_TF_activity</span>(<span class="dt">RNA_tpm =</span> Riaz_data<span class="op">$</span>tpm_RNAseq, </span>
<span id="cb3-13"><a href="#cb3-13"></a>                                   <span class="dt">remove_genes_ICB_proxies =</span> <span class="ot">FALSE</span>)</span>
<span id="cb3-14"><a href="#cb3-14"></a></span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="co"># Computation of ligand-receptor pair weights</span></span>
<span id="cb3-16"><a href="#cb3-16"></a>lrpair_weights &lt;-<span class="st"> </span><span class="kw">compute_LR_pairs</span>(<span class="dt">RNA_tpm =</span> Riaz_data<span class="op">$</span>tpm_RNAseq, </span>
<span id="cb3-17"><a href="#cb3-17"></a>                                   <span class="dt">remove_genes_ICB_proxies =</span> <span class="ot">FALSE</span>, </span>
<span id="cb3-18"><a href="#cb3-18"></a>                                   <span class="dt">cancertype =</span> <span class="st">&quot;pancan&quot;</span>)</span>
<span id="cb3-19"><a href="#cb3-19"></a></span>
<span id="cb3-20"><a href="#cb3-20"></a><span class="co"># Computation of cell-cell interaction scores</span></span>
<span id="cb3-21"><a href="#cb3-21"></a>ccpair_scores &lt;-<span class="st"> </span><span class="kw">compute_CC_pairs_grouped</span>(<span class="dt">lrpairs =</span> lrpair_weights, </span>
<span id="cb3-22"><a href="#cb3-22"></a>                                          <span class="dt">cancertype =</span> <span class="st">&quot;pancan&quot;</span>)</span></code></pre></div>
<p>You can additionally compute published hallmarks of the immune response on your patients’ data.</p>
<table>
<thead>
<tr class="header">
<th>Hallmark of the immune response</th>
<th>Original study</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Cytolytic activity (CYT)</td>
<td>Rooney et al, Cell, 2015</td>
</tr>
<tr class="even">
<td>Roh immune score (Roh_IS)</td>
<td>Roh et al., Sci. Transl. Med., 2017</td>
</tr>
<tr class="odd">
<td>Chemokine signature (chemokines)</td>
<td>Messina et al., Nat. Sci. Rep., 2012</td>
</tr>
<tr class="even">
<td>Davoli immune signature (Davoli_IS)</td>
<td>Davoli et al., Science 2017</td>
</tr>
<tr class="odd">
<td>IFNy signature (IFNy)</td>
<td>Ayers et al., JCI, 2017</td>
</tr>
<tr class="even">
<td>Expanded immune signature (Ayers_expIS)</td>
<td>Ayers et al., JCI, 2017</td>
</tr>
<tr class="odd">
<td>T-cell inflamed signature (Tcell_inflamed)</td>
<td>Ayers et al., JCI, 2017</td>
</tr>
<tr class="even">
<td>Repressed immune resistance (RIR)</td>
<td>Jerby-Arnon et al., Cell, 2018</td>
</tr>
<tr class="odd">
<td>Tertiary lymphoid structures signature (TLS)</td>
<td>Cabrita et al., Nature, 2020</td>
</tr>
</tbody>
</table>
<p>By default all hallmarks of the immune response are used, however you can set your own <code>list_gold_standards</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>tasks &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;CYT&quot;</span>, <span class="st">&quot;Roh_IS&quot;</span>, <span class="st">&quot;chemokines&quot;</span>, <span class="st">&quot;Davoli_IS&quot;</span>, <span class="st">&quot;IFNy&quot;</span>, <span class="st">&quot;Ayers_expIS&quot;</span>, <span class="st">&quot;Tcell_inflamed&quot;</span>, <span class="st">&quot;RIR&quot;</span>, <span class="st">&quot;TLS&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2"></a>immune_response &lt;-<span class="st"> </span><span class="kw">compute_gold_standards</span>(<span class="dt">RNA_tpm =</span> Riaz_data<span class="op">$</span>tpm_RNAseq, </span>
<span id="cb4-3"><a href="#cb4-3"></a>                                          <span class="dt">list_gold_standards =</span> tasks, </span>
<span id="cb4-4"><a href="#cb4-4"></a>                                          <span class="dt">cancertype =</span> <span class="st">&quot;SKCM&quot;</span>,</span>
<span id="cb4-5"><a href="#cb4-5"></a>                                          <span class="dt">output_file_path =</span> <span class="st">&quot;/Users/Oscar/Desktop/Riaz&quot;</span>)</span></code></pre></div>
<p>We are going to use the systems-based signatures computed previously to predict patients’ immune response. <code>predict_immune_response</code> returns compute predictions for single system-based signatures and pairwise combinations of these system-based signatures when <code>include_pairwise_combos = TRUE</code>.</p>
<p>Because models were built in a cancer-type-specific fashion, the user is required to indicate to which <code>cancertype</code> the bulk RNA-seq data belongs.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a></span>
<span id="cb5-2"><a href="#cb5-2"></a>predictions_immune_response &lt;-<span class="st"> </span><span class="kw">predict_immune_response</span>(<span class="dt">pathways =</span> pathway_activity,</span>
<span id="cb5-3"><a href="#cb5-3"></a>                                                       <span class="dt">immunecells =</span> cell_fractions,</span>
<span id="cb5-4"><a href="#cb5-4"></a>                                                       <span class="dt">lrpairs =</span> lrpair_weights,</span>
<span id="cb5-5"><a href="#cb5-5"></a>                                                       <span class="dt">tfs =</span> tf_activity,</span>
<span id="cb5-6"><a href="#cb5-6"></a>                                                       <span class="dt">ccpairs =</span> ccpair_scores,</span>
<span id="cb5-7"><a href="#cb5-7"></a>                                                       <span class="dt">include_pairwise_combos =</span> <span class="ot">FALSE</span>,</span>
<span id="cb5-8"><a href="#cb5-8"></a>                                                       <span class="dt">cancertype =</span> <span class="st">&quot;SKCM&quot;</span>, </span>
<span id="cb5-9"><a href="#cb5-9"></a>                                                       <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>Depending on the patient’s response information available, we can now think of two possible scenarios where:</p>
<ul>
<li><code>real_patient_response</code> is <strong>known</strong> and the performance of the computed predictions can be evaluated. In this case, we expect the user to provide a character <code>vector</code> containing clinical patients’ response (where non-responders are labeled as NR and responders as R).</li>
<li><code>real_patient_response</code> is <strong>unknown</strong> and an score is provided assessing patient-specific likelihood of response to therapy.</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw">assess_immune_response</span>(<span class="dt">predictions_immune_response =</span> predictions_immune_response,</span>
<span id="cb6-2"><a href="#cb6-2"></a>                       <span class="dt">real_patient_response =</span> Riaz_data<span class="op">$</span>patient_response,</span>
<span id="cb6-3"><a href="#cb6-3"></a>                       <span class="dt">RNA_tpm =</span> Riaz_data<span class="op">$</span>tpm_RNAseq,</span>
<span id="cb6-4"><a href="#cb6-4"></a>                       <span class="dt">output_file_path =</span> <span class="st">&quot;/Users/Oscar/Desktop/Riaz&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5"></a>                       <span class="dt">cancertype =</span> <span class="st">&quot;SKCM&quot;</span>)</span></code></pre></div>
<p>We can go further and analyze systems biomarkers present in the user’s bulk RNA-seq data. <code>explore_biomarkers</code> allows to investigate mechanisms behind patients’ response to treatment through interpretable biomarkers.</p>
<p>In order to leverage the biomarkers information obtained during model training, you need to specify again which <code>cancertype</code> the bulk RNA-seq data belongs.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">explore_biomarkers</span>(<span class="dt">pathways =</span> pathways_scores,</span>
<span id="cb7-2"><a href="#cb7-2"></a>                   <span class="dt">immunecells =</span> cell_fractions,</span>
<span id="cb7-3"><a href="#cb7-3"></a>                   <span class="co"># lrpairs = lrpairs_weights,</span></span>
<span id="cb7-4"><a href="#cb7-4"></a>                   <span class="co"># tfs = tfs_scores,</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>                   <span class="co"># ccpairsgroupedscores = ccpairsgrouped_scores,</span></span>
<span id="cb7-6"><a href="#cb7-6"></a>                   <span class="dt">cancertype =</span> cancertype,</span>
<span id="cb7-7"><a href="#cb7-7"></a>                   <span class="dt">real_patient_response =</span> Riaz_data<span class="op">$</span>patient_response,</span>
<span id="cb7-8"><a href="#cb7-8"></a>                   <span class="dt">output_file_path =</span> <span class="st">&quot;/Users/Oscar/Desktop/Riaz&quot;</span>)</span></code></pre></div>
<div id="ready-to-go" class="section level2">
<h2>Ready to go!</h2>
<p>Once you go through the full framework of <code>easier</code>, you are ready to engage in the app of <code>easier</code>. To do so, you can just launch the <code>easier()</code> app:</p>
</div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1>Session info</h1>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw">sessionInfo</span>()</span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co">#&gt; Running under: macOS High Sierra 10.13.6</span></span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="co">#&gt; </span></span>
<span id="cb8-6"><a href="#cb8-6"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb8-7"><a href="#cb8-7"></a><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span></span>
<span id="cb8-8"><a href="#cb8-8"></a><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span></span>
<span id="cb8-9"><a href="#cb8-9"></a><span class="co">#&gt; </span></span>
<span id="cb8-10"><a href="#cb8-10"></a><span class="co">#&gt; locale:</span></span>
<span id="cb8-11"><a href="#cb8-11"></a><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span id="cb8-12"><a href="#cb8-12"></a><span class="co">#&gt; </span></span>
<span id="cb8-13"><a href="#cb8-13"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb8-14"><a href="#cb8-14"></a><span class="co">#&gt; [1] stats4    parallel  stats     graphics  grDevices utils     datasets </span></span>
<span id="cb8-15"><a href="#cb8-15"></a><span class="co">#&gt; [8] methods   base     </span></span>
<span id="cb8-16"><a href="#cb8-16"></a><span class="co">#&gt; </span></span>
<span id="cb8-17"><a href="#cb8-17"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb8-18"><a href="#cb8-18"></a><span class="co">#&gt;  [1] easier_1.0.3                arules_1.6-6               </span></span>
<span id="cb8-19"><a href="#cb8-19"></a><span class="co">#&gt;  [3] Matrix_1.3-0                pdist_1.2                  </span></span>
<span id="cb8-20"><a href="#cb8-20"></a><span class="co">#&gt;  [5] preprocessCore_1.50.0       dorothea_1.0.1             </span></span>
<span id="cb8-21"><a href="#cb8-21"></a><span class="co">#&gt;  [7] ggplot2_3.3.3               ROCR_1.0-11                </span></span>
<span id="cb8-22"><a href="#cb8-22"></a><span class="co">#&gt;  [9] DESeq2_1.28.1               SummarizedExperiment_1.18.2</span></span>
<span id="cb8-23"><a href="#cb8-23"></a><span class="co">#&gt; [11] DelayedArray_0.14.1         matrixStats_0.57.0         </span></span>
<span id="cb8-24"><a href="#cb8-24"></a><span class="co">#&gt; [13] GenomicRanges_1.40.0        GenomeInfoDb_1.24.2        </span></span>
<span id="cb8-25"><a href="#cb8-25"></a><span class="co">#&gt; [15] IRanges_2.22.2              S4Vectors_0.26.1           </span></span>
<span id="cb8-26"><a href="#cb8-26"></a><span class="co">#&gt; [17] viper_1.22.0                Biobase_2.48.0             </span></span>
<span id="cb8-27"><a href="#cb8-27"></a><span class="co">#&gt; [19] BiocGenerics_0.34.0         progeny_1.13.1             </span></span>
<span id="cb8-28"><a href="#cb8-28"></a><span class="co">#&gt; </span></span>
<span id="cb8-29"><a href="#cb8-29"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb8-30"><a href="#cb8-30"></a><span class="co">#&gt;  [1] mixtools_1.2.0         tidyr_1.1.2            bit64_4.0.5           </span></span>
<span id="cb8-31"><a href="#cb8-31"></a><span class="co">#&gt;  [4] splines_4.0.2          blob_1.2.1             GenomeInfoDbData_1.2.3</span></span>
<span id="cb8-32"><a href="#cb8-32"></a><span class="co">#&gt;  [7] remotes_2.2.0          yaml_2.2.1             ggrepel_0.9.0         </span></span>
<span id="cb8-33"><a href="#cb8-33"></a><span class="co">#&gt; [10] pillar_1.4.7           RSQLite_2.2.1          lattice_0.20-41       </span></span>
<span id="cb8-34"><a href="#cb8-34"></a><span class="co">#&gt; [13] glue_1.4.2             digest_0.6.27          RColorBrewer_1.1-2    </span></span>
<span id="cb8-35"><a href="#cb8-35"></a><span class="co">#&gt; [16] XVector_0.28.0         colorspace_2.0-0       htmltools_0.5.0       </span></span>
<span id="cb8-36"><a href="#cb8-36"></a><span class="co">#&gt; [19] XML_3.99-0.5           pkgconfig_2.0.3        genefilter_1.70.0     </span></span>
<span id="cb8-37"><a href="#cb8-37"></a><span class="co">#&gt; [22] zlibbioc_1.34.0        purrr_0.3.4            xtable_1.8-4          </span></span>
<span id="cb8-38"><a href="#cb8-38"></a><span class="co">#&gt; [25] scales_1.1.1           BiocParallel_1.22.0    tibble_3.0.4          </span></span>
<span id="cb8-39"><a href="#cb8-39"></a><span class="co">#&gt; [28] annotate_1.66.0        generics_0.1.0         ellipsis_0.3.1        </span></span>
<span id="cb8-40"><a href="#cb8-40"></a><span class="co">#&gt; [31] withr_2.3.0            survival_3.2-7         magrittr_2.0.1        </span></span>
<span id="cb8-41"><a href="#cb8-41"></a><span class="co">#&gt; [34] crayon_1.3.4           memoise_1.1.0          evaluate_0.14         </span></span>
<span id="cb8-42"><a href="#cb8-42"></a><span class="co">#&gt; [37] MASS_7.3-53            segmented_1.3-1        class_7.3-17          </span></span>
<span id="cb8-43"><a href="#cb8-43"></a><span class="co">#&gt; [40] tools_4.0.2            lifecycle_0.2.0        stringr_1.4.0         </span></span>
<span id="cb8-44"><a href="#cb8-44"></a><span class="co">#&gt; [43] bcellViper_1.24.0      kernlab_0.9-29         locfit_1.5-9.4        </span></span>
<span id="cb8-45"><a href="#cb8-45"></a><span class="co">#&gt; [46] munsell_0.5.0          AnnotationDbi_1.50.3   compiler_4.0.2        </span></span>
<span id="cb8-46"><a href="#cb8-46"></a><span class="co">#&gt; [49] e1071_1.7-4            rlang_0.4.10           grid_4.0.2            </span></span>
<span id="cb8-47"><a href="#cb8-47"></a><span class="co">#&gt; [52] RCurl_1.98-1.2         bitops_1.0-6           rmarkdown_2.6         </span></span>
<span id="cb8-48"><a href="#cb8-48"></a><span class="co">#&gt; [55] gtable_0.3.0           DBI_1.1.0              R6_2.5.0              </span></span>
<span id="cb8-49"><a href="#cb8-49"></a><span class="co">#&gt; [58] gridExtra_2.3          knitr_1.30             dplyr_1.0.2           </span></span>
<span id="cb8-50"><a href="#cb8-50"></a><span class="co">#&gt; [61] bit_4.0.4              KernSmooth_2.23-18     stringi_1.5.3         </span></span>
<span id="cb8-51"><a href="#cb8-51"></a><span class="co">#&gt; [64] Rcpp_1.0.5             vctrs_0.3.6            geneplotter_1.66.0    </span></span>
<span id="cb8-52"><a href="#cb8-52"></a><span class="co">#&gt; [67] tidyselect_1.1.0       xfun_0.19</span></span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
